dcc-submission-loader/src/main/java/org/icgc/dcc/submission/loader/db/postgres/PostgresDatabaseService.java

```diff
@@ -0,0 +1,262 @@
+    val joinedPks = Joiners.COMMA.join(pks);
+
+    val sql = "CREATE INDEX " + indexName + " ON " + tableName + " ( " + joinedPks + " ) ";
+    log.debug("{}", sql);
+
+    jdbcTemplate.execute(sql);
+  }
+
+  private void dropIndex(String indexName) {
+    val sql = "DROP INDEX IF EXISTS " + indexName;
+    jdbcTemplate.execute(sql);
+  }
+
+  private void createTable(String release, String type) {
+    val sqlBuilder = new StringBuilder();
+    val tableName = getTableName(release, type);
+    sqlBuilder.append("CREATE TABLE " + tableName + " (");
+
+    val pks = submissionMetadataService.getPrimaryKey(type);
+    boolean hasPreviousField = false;
+    boolean hasDonorId = false;
+    for (val entry : submissionMetadataService.getFields(type).entrySet()) {
+      val fieldName = entry.getKey();
+      if (hasPreviousField) {
+        sqlBuilder.append(", ");
+      }
+      hasPreviousField = true;
+
+      sqlBuilder.append(fieldName + " varchar(10000)");
+
+      if (pks.contains(fieldName)) {
+        sqlBuilder.append(" NOT NULL");
+      }
+
+      if (fieldName.equals("donor_id")) {
+        hasDonorId = true;
+      }
+    }
+
+    // Add project_id
+    sqlBuilder.append(", " + PROJECT_ID + " varchar(7)");
+    // Add donor_id
+    if (!hasDonorId) {
+      sqlBuilder.append(", " + DONOR_ID_FIELD_NAME + " varchar(500)");
+    }
+
+    sqlBuilder.append(" )");
+
+    val sql = sqlBuilder.toString();
+    log.debug(sql);
+
+    jdbcTemplate.execute(sql);
+  }
+
+  private void populateProjects(String release, Iterable<Project> projects) {
+    log.debug("Populating projects table...");
+    val sql = "INSERT INTO " + getTableName(release, PROJECT_TABLE) + " VALUES (?, ?, ?)";
+    for (val project : projects) {
+      log.debug("{}", project);
+      jdbcTemplate.update(sql, new Object[] { project.getProjectId(), project.getProjectName(),
+          project.getState().getName() });
+    }
+  }
+
+  private String getTableName(String release, String type) {
+    return release.toLowerCase() + DOT + type;
+  }
+
+  private void dropTable(String release, String type) {
+    val tableName = getTableName(release, type);
+    jdbcTemplate.execute("DROP TABLE IF EXISTS " + tableName + " CASCADE");
+  }
+
+  private List<String> getFileTypes(String release) {
+    return submissionMetadataService.getFileTypes().stream()
+        .map(ft -> ft.getType())
+        .collect(toImmutableList());
+  }
+
+  private String createParentChildJoinClause(List<String> childPks, Map<String, String> parentPks) {
+    val joinClause = new StringBuilder();
+
+    boolean hasPrevCondition = false;
+    for (val childPk : childPks) {
+      if (hasPrevCondition) {
+        joinClause.append(" AND ");
+      }
+
+      joinClause.append(format("child.%s = parent.%s", childPk, parentPks.get(childPk)));
+      hasPrevCondition = true;
+    }
+
+    // Add project_id
+    joinClause.append(format(" AND child.%s = parent.%s", PROJECT_ID, PROJECT_ID));
```

Comment:
Might want to experiment with http://www.jooq.org/. I think it could be quite useful.